### PR TITLE
Fix #683 - Extract HTTP client logic into providers

### LINKED ISSUE #683: Adding some abstraction arround http client
I'd like to use scribejava with Okhttp where scribejava will use Okhttp.

I found something going into this direction: #484 and #485 

Is it something you can accept?

About implementation, we can imagine having a module like `scribejava-okhttp` (and some others) which is responsible to integrate scribejava and okhttp together. At least, this integration could be done in a 3rd party project.

### PR DESCRIPTION
**Pull request by juherr**
Proposition for #683

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -17,6 +17,8 @@
     <modules>
         <module>scribejava-core</module>
         <module>scribejava-apis</module>
+        <module>scribejava-ahc</module>
+        <module>scribejava-ning</module>
```
Comment (kullfar) [inline pom.xml]: Sometime ScribeJava will maybe have some new set of submodules.
Let this httpclient submodules pack will be named
scribejava-httpclient-ahc
and
scribejava-httpclient-ning

File: scribejava-ahc/src/main/java/com/github/scribejava/core/async/ahc/AhcHttpClient.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package com.github.scribejava.core.async.ahc;
```
Comment (kullfar) [inline scribejava-ahc/src/main/java/com/github/scribejava/core/async/ahc/AhcHttpClient.java]: It's not in 'core' module now. And I think, "async" is not the valuable keyword already.
So as for me the package name should be just
com.github.scribejava.httpclient.ahc
to reflect maven arifact id

**Comment by juherr** [comment]
@kullfar Your remarks are not a problem. Just say me when your review is done and I'll make updates.

File: scribejava-core/src/main/java/com/github/scribejava/core/async/HttpProvider.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.github.scribejava.core.async;
+
+import com.github.scribejava.core.model.HttpClient;
+
+public interface HttpProvider {
```
Comment (kullfar) [inline scribejava-core/src/main/java/com/github/scribejava/core/async/HttpProvider.java]: rename to HttpClientProvider pls
to reflect its client nature

File: scribejava-ahc/src/main/java/com/github/scribejava/core/async/ahc/AhcProvider.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+public class AhcProvider implements HttpProvider {
+
+    @Override
+    public HttpClient createClient(HttpClient.Config config) {
+        if (!(config instanceof AhcHttpClientConfig)) {
+            return null;
+        }
+        final AhcHttpClientConfig httpClientConfig = (AhcHttpClientConfig) config;
+        return new AhcHttpClient(httpClientConfig);
```
Comment (kullfar) [inline scribejava-ahc/src/main/java/com/github/scribejava/core/async/ahc/AhcProvider.java]: just a bit more readable
```
        if (config instanceof AhcHttpClientConfig) {
            return new AhcHttpClient((AhcHttpClientConfig) config);
        }
        return null;
```
IMHO

File: scribejava-ahc/src/main/resources/META-INF/services/com.github.scribejava.core.async.HttpProvider [hunk 5]
```diff
@@ -0,0 +1 @@
+com.github.scribejava.core.async.ahc.AhcProvider
```
Comment (kullfar) [inline scribejava-ahc/src/main/resources/META-INF/services/com.github.scribejava.core.async.HttpProvider]: "No newline at end of file."
It's common to add newline at all files, let it be so, thx

File: scribejava-core/src/main/java/com/github/scribejava/core/builder/ServiceBuilder.java [hunk 6]
```diff
@@ -28,11 +30,7 @@
     private Integer readTimeout;
 
     //async version only
```
Comment (kullfar) [inline scribejava-core/src/main/java/com/github/scribejava/core/builder/ServiceBuilder.java]: not relevant comment I thinks
smth like this
//not-default httpclient only

**Comment by kullfar** [comment]
Great job!
Looks like the ideal, right what I had in mind, thx a lot!
I left some minor comments.

**Comment by juherr** [comment]
Updated according to your remarks.

**Comment by kullfar** [comment]
I tested the branch, all is awesome, merged with some small additions https://github.com/scribejava/scribejava/commit/8c2f85bcc269e85c9609ad78a65bf5ebcc7d4ebe

**Comment by kullfar** [comment]
Thx again

**Pull request merged**